library/Vanilla/Models/ModelCache.php

```diff
@@ -0,0 +1,190 @@
+<?php
+/**
+ * @author Adam Charron <adam.c@vanillaforums.com>
+ * @copyright 2009-2020 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+namespace Vanilla\Models;
+
+use Vanilla\FeatureFlagHelper;
+use Vanilla\InjectableInterface;
+
+/**
+ * Cache for records out of various models.
+ *
+ * Particularly good support for PipelineModel.
+ */
+class ModelCache implements InjectableInterface {
+
+    /** @var int When we hit this size of incrementing key, we reset from 0. */
+    const MAX_INCREMENTING_KEY = 1000000;
+
+    /** @var string */
+    const INCREMENTING_KEY_NAMESPACE = 'vanillaIncrementingKey';
+
+    const GLOBAL_DEFAULT_OPTIONS = [
+        \Gdn_Cache::FEATURE_EXPIRY => 600,
+    ];
+
+    const DISABLE_FEATURE_FLAG = "Feature.DisableNewModelCaching.Enabled";
+
+    /** @var string */
+    private $namespace;
+
+    /** @var number */
+    private $defaultCacheOptions;
+
+    /** @var int */
+    private $incrementingKey;
+
+    /** @var \Gdn_Cache */
+    private $cache;
+
+    /** @var bool */
+    private $isFeatureDisabled;
+
+    /**
+     * Create a cache from a model.
+     *
+     * @param Model $model The model to create from.
+     * @param array|null $defaultCacheOptions Default options to apply for storing cache values.
+     *
+     * @return ModelCache
+     */
+    public static function fromModel(Model $model, ?array $defaultCacheOptions = []): ModelCache {
+        $cache = \Gdn::getContainer()->getArgs(ModelCache::class, [$model->getTable(), $defaultCacheOptions]);
+        return $cache;
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param string $namespace
+     * @param array|null $defaultCacheOptions Default options to apply for storing cache values.
+     */
+    public function __construct(string $namespace, ?array $defaultCacheOptions = []) {
+        $this->namespace = $namespace;
+        $this->defaultCacheOptions = array_merge(self::GLOBAL_DEFAULT_OPTIONS, $defaultCacheOptions ?? []);
+        $this->isFeatureDisabled = \Gdn::config(self::DISABLE_FEATURE_FLAG);
```

Comment:
I'm okay with a global call since this is just a feature flag. However, please use the `FeatureFlagHelper` methods.

Reply:
Done.